src/xdp/offload.c

```diff
@@ -362,10 +362,86 @@ XdpIrpInterfaceOffloadQeoSet(
     _In_ IO_STACK_LOCATION *IrpSp
     )
 {
-    UNREFERENCED_PARAMETER(InterfaceObject);
-    UNREFERENCED_PARAMETER(Irp);
-    UNREFERENCED_PARAMETER(IrpSp);
-    return STATUS_NOT_IMPLEMENTED;
+    NTSTATUS Status;
+    XDP_OFFLOAD_PARAMS_QEO QeoParams = {0};
+    const XDP_QUIC_CONNECTION *ConnectionsIn = Irp->AssociatedIrp.SystemBuffer;
+    UINT32 InputBufferLength = IrpSp->Parameters.DeviceIoControl.InputBufferLength;
+    UINT32 OutputBufferLength = IrpSp->Parameters.DeviceIoControl.InputBufferLength;
+    UINT32 BytesWritten = 0;
+
+    TraceEnter(TRACE_CORE, "Interface=%p", InterfaceObject);
+
+    if (InputBufferLength == 0) {
+        Status = STATUS_INVALID_PARAMETER;
+        goto Exit;
+    }
+
+    QeoParams.Connections = ConnectionsIn;
+    QeoParams.ConnectionsSize = InputBufferLength;
+    QeoParams.ConnectionCount = 0;
+
+    while (InputBufferLength > 0) {
+        //
+        // Validate input.
+        //
+
+        if (InputBufferLength < sizeof(ConnectionsIn->Header) ||
+            InputBufferLength < ConnectionsIn->Header.Size) {
+            TraceError(
+                TRACE_CORE,
+                "Interface=%p Input buffer length too small InputBufferLength=%u",
+                InterfaceObject, InputBufferLength);
+            Status = STATUS_INVALID_PARAMETER;
+            goto Exit;
+        }
+
+        if (ConnectionsIn->Header.Revision != XDP_QUIC_CONNECTION_REVISION_1 ||
+            ConnectionsIn->Header.Size < XDP_SIZEOF_QUIC_CONNECTION_REVISION_1) {
+            TraceError(
+                TRACE_CORE, "Interface=%p Unsupported revision Revision=%u Size=%u",
+                InterfaceObject, ConnectionsIn->Header.Revision, ConnectionsIn->Header.Size);
+            Status = STATUS_INVALID_PARAMETER;
+            goto Exit;
+        }
+
+        if ((UINT32)ConnectionsIn->Operation > (UINT32)XDP_QUIC_OPERATION_REMOVE ||
+            (UINT32)ConnectionsIn->Direction > (UINT32)XDP_QUIC_DIRECTION_RECEIVE ||
+            (UINT32)ConnectionsIn->DecryptFailureAction > (UINT32)XDP_QUIC_DECRYPT_FAILURE_ACTION_CONTINUE ||
+            ConnectionsIn->RESERVED != 0 ||
+            (UINT32)ConnectionsIn->CipherType > (UINT32)XDP_QUIC_CIPHER_TYPE_AEAD_AES_128_CCM ||
+            (UINT32)ConnectionsIn->AddressFamily > (UINT32)XDP_QUIC_ADDRESS_FAMILY_INET6 ||
+            ConnectionsIn->ConnectionIdLength > sizeof(ConnectionsIn->ConnectionId)) {
+            Status = STATUS_INVALID_PARAMETER;
+            goto Exit;
+        }
```

Comment:
Do we _need_ to do these kinds of checks, or can we simply be a pass through. Additionally, for `RESERVED` fields the logic generally is "always set to 0, but never read" in case it changes outside of your knowledge in the future. I think we should design XDP to require as few possible future changes as possible. If a new cipher type is added in the future, I don't want to require a new version of XDP. Now, I'm totally fine for the test miniport to have these kinds of checks though.

Reply:
Ah, I misunderstood the meaning of "never read" - yes that check should be skipped. In general, though, I am very strongly against allowing user mode to send junk down to an NDIS driver. We are writing a stack bypass, but that should not go so far as to allow arbitrary IO to the entire ecosystem of Windows drivers.